llama_index/core/base_retriever.py

```diff
@@ -39,6 +52,122 @@ def _get_prompt_modules(self) -> PromptMixinType:
+            ]
+        elif isinstance(obj, BaseRetriever):
+            return obj.retrieve(query_bundle)
+        elif isinstance(obj, QueryComponent):
+            component_keys = obj.input_keys.required_keys
+            if len(component_keys) > 1:
+                raise ValueError(
+                    f"QueryComponent {obj} has more than one input key: {component_keys}"
+                )
+            elif len(component_keys) == 0:
+                component_response = obj.run_component()
+            else:
+                kwargs = {next(iter(component_keys)): query_bundle.query_str}
+                component_response = obj.run_component(**kwargs)
+
+            result_output = str(next(iter(component_response.values())))
+            return [NodeWithScore(node=TextNode(text=result_output), score=score)]
+        else:
+            raise ValueError(f"Object {obj} is not retrievable.")
+
+    async def _aretrieve_from_object(
+        self,
+        obj: Any,
+        query_bundle: QueryBundle,
+        score: float,
+    ) -> List[NodeWithScore]:
+        """Retrieve nodes from object."""
+        if isinstance(obj, NodeWithScore):
+            return [obj]
+        elif isinstance(obj, BaseNode):
+            return [NodeWithScore(node=obj, score=score)]
+        elif isinstance(obj, BaseQueryEngine):
+            response = await obj.aquery(query_bundle)
+            return [NodeWithScore(node=TextNode(text=str(response)), score=score)]
+        elif isinstance(obj, BaseRetriever):
+            return await obj.aretrieve(query_bundle)
+        elif isinstance(obj, QueryComponent):
+            component_keys = obj.input_keys.required_keys
+            if len(component_keys) > 1:
+                raise ValueError(
+                    f"QueryComponent {obj} has more than one input key: {component_keys}"
+                )
+            elif len(component_keys) == 0:
+                component_response = await obj.arun_component()
+            else:
+                kwargs = {next(iter(component_keys)): query_bundle.query_str}
+                component_response = await obj.arun_component(**kwargs)
+
+            result_output = str(next(iter(component_response.values())))
+            return [NodeWithScore(node=TextNode(text=result_output), score=score)]
+        else:
+            raise ValueError(f"Object {obj} is not retrievable.")
+
+    def _handle_recursive_retrieval(
+        self, query_bundle: QueryBundle, nodes: List[NodeWithScore]
+    ) -> List[NodeWithScore]:
+        retrieved_nodes: List[NodeWithScore] = []
+        for n in nodes:
+            node = n.node
+            score = n.score or 1.0
+            if isinstance(node, IndexNode):
+                obj = self.object_map.get(node.index_id, None)
+                if obj is not None:
+                    retrieved_nodes.extend(
+                        self._retrieve_from_object(
+                            obj, query_bundle=query_bundle, score=score
```

Comment:
if we are getting different nodes from different index, the score range could be different.
we might need to normalize score to make it comparable

Reply:
eh, lets leave that for the future. The user can always handle normalizing the retrieved nodes, and opening the door to adding some fusion/rerank function will get very complicated